// @dp/knb/env.js
function getENV() {
  // make sure the script not throwing error under nodejs.
  var win  = typeof window !== 'undefined'    ? window              : {};
  var href = typeof location !== 'undefined'  ? location.href       : '';
  var ua   = typeof navigator !== 'undefined' ? navigator.userAgent : '';
  var is_stupid_ios = /ios|iphone|ipad|ipod/i.test(ua); // iOS下UA非常不可靠

  var regDP = win._KNB_REGDP || /dp\/com\.dianping\.[\w\.]+\/([\d\.]+)/i;
  var regMT = win._KNB_REGMT || /moviepro|MTNB|Meituan|MOMA/i;
  var regHB = win._KNB_REGHB || /iHotel|HBNB/i;
  var regWX = win._KNB_REGWX || /micromessenger/i;
  var regTitans = win._KNB_REGTITANS || /TitansX/i;

  var isWX     = regWX.test(ua);
  var isDPApp  = regDP.test(ua);
  var isMTNB   = regMT.test(ua) || /f=(iphone|ipad|ipod|android)/.test(href);
  var isHBNB   = regHB.test(ua) && !win._MTNB;
  var isTitans = function() {
    if (href.indexOf('knb_force_use_titans') > -1) return true;
    if (is_stupid_ios) return !!win._TitansX;
    return regTitans.test(ua);
  }();
  // order matters
  if (isWX)     { isTitans = isHBNB = isDPApp = isMTNB = false; }
  if (isTitans) { isDPApp  = isWX   = isMTNB  = isHBNB = false; }
  if (isHBNB)   { isTitans = isWX   = isDPApp = isMTNB = false; }
  if (isMTNB)   { isTitans = isWX   = isDPApp = isHBNB = false; }
  if (isDPApp)  { isTitans = isWX   = isMTNB  = isHBNB = false; }

  return {
    isWX: isWX,
    isTitans: isTitans,
    isHBNB: isHBNB,
    isDPApp: isDPApp,
    isMTNB: isMTNB,
  };
}

module.exports = getENV();
module.exports.getENV = getENV;  // make it testable
