// @dp/knb/lib/titans/api/subscribe.js
var pageEvents = ['appear', 'disappear'];
var count = 0;
var getRandomId = () => ('sub' + (++count))
var noop = function() {}

module.exports = {
  message: 'subscribe',
  mapper: function({action, handle=noop, success=noop, fail=noop}) {

    this._eventQueue = this._eventQueue || {};
    this._subscriptionMap = this._subscriptionMap || {};
    var subId = getRandomId();

    this._subscriptionMap[subId] = {action, handle};

    var _events = this._eventQueue;
    var eventName = action;

    var eventHandler = function(e) {
      (_events[eventName] || []).filter(Boolean).forEach(fn => fn(e));
    };

    function registerPageEvents() {
      if (!eventName) return;
      var capital = str => str.slice(0,1).toUpperCase() + str.slice(1);
      if (window.DPApp) window.DPApp['on' + capital(eventName)] = eventHandler;
      success({subId});
    }

    if (_events[eventName]) {
      //事件只需要向客户端注册一次，事件队列有JS段维护
      success({subId});
      _events[eventName].push(handle);
      return;
    }

    _events[eventName] = [handle];

    //页面级别事件 appear/disappear 直接注册在DPApp全局对象下，事件触发时客户端会直接调用
    if (pageEvents.indexOf(eventName) !== -1) return registerPageEvents()
    return {
      action: eventName,
      success: () => success({subId}),
      handle: eventHandler,
      fail
    };
  }
};
