// @dp/knb/compiled/titans/dp/9.2.0.js
var mixin = require('../../common/mixin');
var util = require('../util/utility');

function noop() {}

module.exports = [
  {
    version: '9.2.0',
    message: 'getUserInfo'
  },
  {
    version: '9.2.0',
    message: 'login',
    mapper: function (opts) {
      var self = this;
      var success = opts.success || noop
      return mixin({}, opts, {
        success: function () {
          self.getUserInfo({
            success: function (ret) {
              success(mixin({}, ret, {
                type: 'dp',
                token: ret.token || ret.userToken,
                uuid: ret.uuid
              }));
            }
          })
        }
      });
    }
  },
  {
    version: '9.2.0',
    message: 'share',
    statics: {
      WECHAT_FRIENDS: 0,
      WECHAT_TIMELINE: 1,
      QQ: 2,
      SMS: 3,
      WEIBO: 4,
      QZONE: 5,
      EMAIL: 6,
      COPY: 7
    },
    mapper: function (opts) {
      opts = mixin({}, opts);
      opts.feed = opts.channel;
      if (!(opts.feed instanceof Array)) {
        opts.feed = [opts.feed];
      }
      if (opts.feed && opts.feed.length === 1) {
        opts.shareType = opts.feed[0];
      }


      return mixin(opts, {
        feed: util.parseFeed(opts.feed),
        url: util.tidyUrlParams(opts.url)
      });
    }
  }
];
