// @hfe/hbnb/_demo/webpack.config.babel.js
'use strict';

import webpack from 'webpack';

export default {
    context: __dirname,
    entry: {
        index: [
            'sanitize.css/sanitize.css',
            'purecss/build/buttons.css',
            'purecss/build/forms-nr.css',
            './index.jsx'
        ]
    },
    output: {
        path: __dirname,
        filename: './[name].js'
    },
    module: {
        loaders: [{
            test: /\.jsx?$/,
            exclude: /node_modules/,
            loader: 'babel',
            query: {
                presets: ['react', 'es2015', 'stage-0'],
                env: {
                    "plugins": ["react-transform"],
                    "extra": {
                        "react-transform": {
                            "transform": [{
                                "transform": "react-transform-hmr",
                                "imports": ["react"],
                                "locals": ["module"]
                            }, {
                                "transform": "react-transform-catch-errors",
                                "imports": [
                                    "react",
                                    "./_demo/index.jsx"
                                ]
                            }]
                        }
                    }
                }
            }
        }, {
            test: /\.css$/,
            loaders: [
                'style',
                'css?sourceMap'
            ]
        }, {
            test: /\.scss$/,
            loaders: [
                'style',
                'css?sourceMap',
                'sass'
            ]
        }]
    },
    plugins: [
        new webpack.ProvidePlugin({
            'Promise': 'pinkie'
        }),
        new webpack.EnvironmentPlugin('NODE_ENV'),
        new webpack.HotModuleReplacementPlugin()
    ],
    devtool: '#inline-source-map',
    debug: true,
    devServer: {
        host: '0.0.0.0',
        contentBase: __dirname,
        inline: true,
        hot: true
    },
    node: {
        process: false,
        setImmediate: false
    }
};
